library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedJoinOperator.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.apex.malhar.lib.window.WindowOption;
+import org.apache.apex.malhar.lib.window.WindowState;
+import org.apache.apex.malhar.lib.window.WindowedStorage;
+
+import com.google.common.base.Function;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.annotation.InputPortFieldAnnotation;
+
+
+/**
+ * Abstract Windowed Join Operator.
+ */
+public abstract class AbstractWindowedJoinOperator<InputT1, InputT2, OutputT, DataStorageT extends WindowedStorage,
+    RetractionStorageT extends WindowedStorage, AccumulationT extends
+    JoinAccumulation>
+    extends AbstractWindowedOperator<InputT1, OutputT, DataStorageT, RetractionStorageT, AccumulationT>
+    implements JoinWindowedOperator<InputT1, InputT2>
+{
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(AbstractWindowedJoinOperator.class);
+  private Function<InputT2, Long> timestampExtractor2;
+
+  private long latestWatermark1 = -1;  // latest watermark from stream 1
+  private long latestWatermark2 = -1;  // latest watermark from stream 2
+
+
+  private static final transient Collection<? extends Window> GLOBAL_WINDOW_SINGLETON_SET = Collections.singleton(Window.GlobalWindow.getInstance());
+
+  public final transient DefaultInputPort<Tuple<InputT2>> input2 = new DefaultInputPort<Tuple<InputT2>>()
+  {
+    @Override
+    public void process(Tuple<InputT2> tuple)
+    {
+      processTuple2(tuple);
+    }
+  };
+
+  // TODO: This port should be removed when Apex Core has native support for custom control tuples
+  @InputPortFieldAnnotation(optional = true)
+  public final transient DefaultInputPort<ControlTuple> controlInput2 = new DefaultInputPort<ControlTuple>()
+  {
+    @Override
+    public void process(ControlTuple tuple)
+    {
+      if (tuple instanceof ControlTuple.Watermark) {
+        processWatermark2((ControlTuple.Watermark)tuple);
+      }
+    }
+  };
+
+  public void processTuple2(Tuple<InputT2> tuple)
+  {
+    long timestamp = extractTimestamp2(tuple);
+    if (isTooLate(timestamp)) {
+      dropTuple2(tuple);
+    } else {
+      Tuple.WindowedTuple<InputT2> windowedTuple = getWindowedValue2(tuple);
+      // do the accumulation
+      accumulateTuple2(windowedTuple);
+
+      for (Window window : windowedTuple.getWindows()) {
+        WindowState windowState = windowStateMap.get(window);
+        windowState.tupleCount++;
+        // process any count based triggers
+        if (windowState.watermarkArrivalTime == -1) {
+          // watermark has not arrived yet, check for early count based trigger
+          if (earlyTriggerCount > 0 && (windowState.tupleCount % earlyTriggerCount) == 0) {
+            fireTrigger(window, windowState);
+          }
+        } else {
+          // watermark has arrived, check for late count based trigger
+          if (lateTriggerCount > 0 && (windowState.tupleCount % lateTriggerCount) == 0) {
+            fireTrigger(window, windowState);
+          }
```

Comment:
The AbstractWindowedOperator should probably be changed so that you can reuse the code here. Copy-paste code is generally not a good idea.